### ISSUE TITLE
Matching training statistics

### ISSUE THREAD

**Issue by JonathanCMitchell** [MadryLab/cifar10_challenge#16]
I tried to match the training scheme of this network and I was unable to do so using what seems to be the same parameters. 

I pull a random batch for each epoch of size `batch_size` = 48.

Epochs 0 - 40,000 LR = 0.1
Epochs 40,000 - 60,000 LR = 0.01
Epochs 60,000 - 80,000 LR = 0.001
In this case it nearly matches the procedure shown because I have an epoch for each batch, so it takes ~937 epochs to cycle once through my training dataset.

However, the cross entropy loss on my network is near 0 by the time I move past 60k epochs, and the network is only trained on the adversarial samples.

(1) Are the momentum parameters saved after the learning rate is updated? Because it seems like a brand new optimizer is being created after the 40th epoch, and after the 60th epoch. 

(2) Did you experience anything like this? I am using an off the shelf WideResNet 30 from [here](https://github.com/meliketoy/wide-resnet.pytorch/blob/master/networks/wide_resnet.py).

My training set is 45k images, and I have a validation set of 5k images. Each adversarial sample is computed using the training set. I am able to get ~100% accuracy on the natural images and around 100% on the adversarial trained images but only **80% on the natural testing images.** 

![xent_adv_train_nat](https://user-images.githubusercontent.com/13068956/67813094-f3298d80-fa5d-11e9-9166-90e22d41cefd.png)
![acc_adv_train_nat](https://user-images.githubusercontent.com/13068956/67813095-f3298d80-fa5d-11e9-8e09-21bcc7ae993e.png)

**Comment by dtsip** [comment]
> However, the cross entropy loss on my network is near 0 by the time I move past 60k epochs, and the network is only trained on the adversarial samples.

There is nothing wrong with that. The model is wide enough that it can fit the training set perfectly.

>(1) Are the momentum parameters saved after the learning rate is updated? Because it seems like a brand new optimizer is being created after the 40th epoch, and after the 60th epoch.

We are not creating a new optimizer at each drop. We just provide an LR schedule to momentumOptimizer directly.

>(2) Did you experience anything like this? I am using an off the shelf WideResNet 30 from here.

Yes, this behavior is roughly consistent with what we experience. What is the exact phenomenon you are referring to?

>My training set is 45k images, and I have a validation set of 5k images. Each adversarial sample is computed using the training set. I am able to get ~100% accuracy on the natural images and around 100% on the adversarial trained images but only 80% on the natural testing images.

Hm, natural test accuracy is typically around 87%. Not sure what is causing this drop. One thing I notice is that we use a batch size of 128. So 40k steps correspond to roughly double the epochs with an LR of 0.1. Not sure if this makes a difference.

**Comment by dtsip** [comment]
Closing for now, feel free to reopen if you need to.
